libs/ml/include/ml/model/model.hpp

```diff
@@ -256,6 +267,48 @@ void Model<TensorType>::UpdateConfig(ModelConfig<DataType> &model_config)
   model_config_ = model_config;
 }
 
+template <typename TensorType>
+std::string Model<TensorType>::InputName()
+{
+  return input_;
+}
+
+template <typename TensorType>
+std::string Model<TensorType>::LabelName()
+{
+  return label_;
+}
+
+template <typename TensorType>
+std::string Model<TensorType>::OutputName()
+{
+  return output_;
+}
+
+template <typename TensorType>
+std::string Model<TensorType>::ErrorName()
+{
+  return error_;
+}
+
+template <typename TensorType>
+typename Model<TensorType>::DataLoaderType *Model<TensorType>::GetDataloader()
+{
+  return dataloader_ptr_.get();
+}
+
+template <typename TensorType>
+typename Model<TensorType>::ModelOptimiserType *Model<TensorType>::GetOptimiser()
+{
+  return optimiser_ptr_.get();
+}
+
+template <typename TensorType>
+typename Model<TensorType>::GraphType *Model<TensorType>::GetGraph()
+{
+  return graph_ptr_.get();
+}
```

Comment:
I'm sure there's a reason for it, but why do these functions return raw pointers?

Reply:
because the model actually OWNS (via a unique ptr) its graph, dataloader, and optimiser. This is good for ordinary use cases (because we shouldn't mess with model internals without a good reason); however distributed ML is a special case where we probably do need to mess with these internals. however, we can't just std::move them out of the model (because this would break the model).

As a compromise here I've exported raw pointers so we can use these internals; but this should indicate to any developer that doing this implies you know exactly what you're doing!

I don't really like this solution much, but I didn't like the alternatives either. Here are some for example:
1. rewrite distributed client functions to only interact with the model (however, this probably requires adding extra functions to model that only exist for this one use case),
2. change models internals to shared_ptrs (but this implies it's fine to extract and manipulate model internals, which it probably isn't)
3. make distributed client a friend of model and use the unique_ptrs directly (perhaps this is a better option?)

Perhaps I will rewrite this as a friend and see if it's cleaner

Reply:
i see that we have already made distributed client a friend of graph, so we might as well do the same for model; then we can ditch those raw pointer accessors.
nice catch :-)

Reply:
I think another solution would be to return a const ref to a unique pointer? But making model a friend of graph sounds better. Unless that would create a circular dependency?